File: boj/graph/2146.java
import java.io.*;
import java.util.*;

class Position {
    int x, y;

    Position(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

public class Main {

    static int[] toRow = {1, -1, 0, 0}, toCol = {0, 0, 1, -1};
    static int N;
    static int[][] map, cost;
    static Queue<Position> que;
    static boolean[][] visited;
    static int result;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        // 초기화
        N = Integer.parseInt(br.readLine());
        map = new int[N][N];
        cost = new int[N][N];
        que = new LinkedList<>();
        visited = new boolean[N][N];
        result = N * N + 1;


        // 입력
        for (int i = 0; i < N; i++) {
            String[] input = br.readLine().split(" ");
            for (int j = 0; j < N; j++) {
                map[i][j] = Integer.parseInt(input[j]);
                if (map[i][j] != 0) que.add(new Position(i, j));
            }
        }

        // 섬의 값을 할당하기
        int value = 1;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (map[i][j] == 0) continue;
                if (visited[i][j]) continue;
                changeValueOfLand(i, j, value);
                value++;
            }
        }

        // 다리 놓는 거리 찾기
        find();

        bw.write(result);
        bw.flush();
    }

    // 섬의 값을 할당하는 함수
    static void changeValueOfLand(int x, int y, int value) {
        Queue<Position> que = new LinkedList<>();

        que.add(new Position(x, y));
        visited[x][y] = true;

        while (!que.isEmpty()) {
            Position cur = que.poll();

            map[cur.x][cur.y] = value;

            for (int i = 0; i < 4; i++) {
                int nextRow = toRow[i] + cur.x;
                int nextCol = toCol[i] + cur.y;


                if (!isInRange(nextRow, nextCol)) continue;
                if (visited[nextRow][nextCol]) continue;
                if (map[nextRow][nextCol] != 0) {
                    que.add(new Position(nextRow, nextCol));
                    visited[nextRow][nextCol] = true;
                }
            }
        }

    }

    // 다리를 찾는 함수
    static void find() {

        boolean[][] visited = new boolean[N][N];

        while (!que.isEmpty()) {
            Position cur = que.poll();

            for (int i = 0; i < 4; i++) {
                int nextRow = toRow[i] + cur.x;
                int nextCol = toCol[i] + cur.y;

                if (!isInRange(nextRow, nextCol)) continue;
                if (visited[nextRow][nextCol]) {
                    // 이미 방문한 적이 있음
                    if (map[cur.x][cur.y] != map[nextRow][nextCol]) {
                        // 현재 섬과 다음 섬이 다르고, 다음 섬의 가중치가 있는 경우==>다리를 놓을 수 있다.
                        result = Math.min(cost[nextRow][nextCol] + cost[cur.x][cur.y], result);
                    }
                }
                else {
                    if (map[nextRow][nextCol] == 0) {
                        // 섬도 없고 다리도 없으면, 다리를 놓는다.
                        map[nextRow][nextCol] = map[cur.x][cur.y]; // 섬을 확장시킨다.
                        cost[nextRow][nextCol] = cost[cur.x][cur.y] + 1; // 다리의 길이
                        visited[nextRow][nextCol] = true; // 방문체크
                        que.add(new Position(nextRow, nextCol));
                    }
                }
            }
        }
    }


    static boolean isInRange(int x, int y) {
        if (x < 0 || y < 0 || x >= N || y >= N) return false;
        return true;
    }
}

File: boj/graph/2146.swift
import Foundation

struct Pos {
    var row: Int, col: Int
    var cnt: Int
    
    init(_ row: Int, _ col: Int) {
        self.row = row
        self.col = col
        self.cnt = 0
    }
    
    init(_ row: Int, _ col: Int, _ cnt: Int) {
        self.row = row
        self.col = col
        self.cnt = cnt
    }
}

guard let inputOfN = readLine() else { exit(0) }
guard let N = Int(inputOfN) else { exit(0) }

var map = [[Int]](repeating: [Int](repeating: 0, count: N), count: N) // 입력 맵
var coloredMap = [[Int]](repeating: [Int](repeating: 0, count: N), count: N) // 섬 판별 맵
var mapOfLength = [[Int]](repeating: [Int](repeating: 0, count: N), count: N) // 다리 길이 맵

var result = Int.max // 결과 값

for i in 0..<N {
    guard let inputOfRow = readLine()?.split(separator: " ").compactMap({Int(String($0))}) else { exit(0) }
    for j in 0..<N {
        map[i][j] = inputOfRow[j]
    }
}


var color = 0
for i in 0..<N {
    for j in 0..<N {
        if map[i][j] == 1 && coloredMap[i][j] == 0 {
            color += 1
            allocateToLand(i, j, color)
        }
    }
}
search()
print(result)

let toward = [[1, 0], [-1, 0], [0, 1], [0, -1]]
func isInRange(_ row: Int, _ col: Int, _ N: Int) -> Bool {
    if row < 0 || col < 0 || row >= N || col >= N { return false }
    return true
}


func allocateToLand(_ row: Int, _ col: Int, _ color: Int) {
    var que = [Pos]()
    que.append(Pos(row, col))
    coloredMap[row][col] = color
    
    while !que.isEmpty {
        let cur = que.removeFirst()
        
        for i in 0..<4 {
            let nextRow = toward[i][0] + cur.row
            let nextCol = toward[i][1] + cur.col
            
            if !isInRange(nextRow, nextCol, N) { continue }
            if map[nextRow][nextCol] == 0 { continue }
            if coloredMap[nextRow][nextCol] == color { continue }
            
            coloredMap[nextRow][nextCol] = color
            que.append(Pos(nextRow, nextCol))
        }
    }
}

func search() {
    
    var que = [Pos]()
    var visited = [[Bool]](repeating: [Bool](repeating: false, count: N), count: N)
    
    for i in 0..<N {
        for j in 0..<N {
            if coloredMap[i][j] > 0 {
                que.append(Pos(i, j, 0))
                visited[i][j] = true
            }
        }
    }
    
    while !que.isEmpty {
        let cur = que.removeFirst()
        for i in 0..<4 {
            let nextRow = cur.row + toward[i][0]
            let nextCol = cur.col + toward[i][1]
            
            if !isInRange(nextRow, nextCol, N) { continue }
            if coloredMap[nextRow][nextCol] > 0 { // 이미 방문했던 곳을 만남
                if coloredMap[nextRow][nextCol] == coloredMap[cur.row][cur.col] { // 내 섬이면, 패스
                    continue
                }
                else { // 내 섬이 아니면 다른 섬에서 오는 다리를 만난 거니까 출력하고 끝냄
                    result = min(result, mapOfLength[cur.row][cur.col] + mapOfLength[nextRow][nextCol])
                }
            }
            else {
                // 새롭게 방문한 곳이라면 큐에 값을 넣고 다리 연결하기
                que.append(Pos(nextRow, nextCol, cur.cnt+1))
                coloredMap[nextRow][nextCol] = coloredMap[cur.row][cur.col]
                mapOfLength[nextRow][nextCol] = mapOfLength[cur.row][cur.col]+1
            }
        }
        
    }
}
